### PR TITLE
RM-7380: Add MethodImplOptions.AggressiveInlining to simple argument checks

### PR DESCRIPTION
**Pull request by iam-mholle**
https://re-motion.atlassian.net/browse/RM-7380

### CODE REVIEW

File: SharedSource/Core/Utilities/ArgumentUtility.cs [hunk 1]
```diff
@@ -53,6 +54,7 @@ namespace Remotion.Utilities
   /// </remarks>
   static partial class ArgumentUtility
   {
+    [MethodImpl (MethodImplOptions.AggressiveInlining)]
```
Comment (MichaelKetting) [inline SharedSource/Core/Utilities/ArgumentUtility.cs:57]: please move to after assertionmethod atrribute

**Comment by iam-mholle** [comment]
- `CheckNotNull`
```text
.method public hidebysig static 
	!!T CheckNotNull<T> (
		string argumentName,
		!!T actualValue
	) cil managed aggressiveinlining 
{
	.custom instance void JetBrains.Annotations.AssertionMethodAttribute::.ctor() = (
		01 00 00 00
	)
	.param [1]
		.custom instance void JetBrains.Annotations.InvokerParameterNameAttribute::.ctor() = (
			01 00 00 00
		)
	.param [2]
		.custom instance void JetBrains.Annotations.AssertionConditionAttribute::.ctor(valuetype JetBrains.Annotations.AssertionConditionType) = (
			01 00 03 00 00 00 00 00
		)
		.custom instance void JetBrains.Annotations.NoEnumerationAttribute::.ctor() = (
			01 00 00 00
		)
	// Method begins at RVA 0x9034
	// Code size 17 (0x11)
	.maxstack 8

	IL_0000: ldarg.1
	IL_0001: box !!T
	IL_0006: brtrue.s IL_000f

	IL_0008: ldarg.0
	IL_0009: newobj instance void [mscorlib]System.ArgumentNullException::.ctor(string)
	IL_000e: throw

	IL_000f: ldarg.1
	IL_0010: ret
} // end of method ArgumentUtility::CheckNotNull
```

- `CheckNotNullOrEmpty`
```text
.method public hidebysig static 
	string CheckNotNullOrEmpty (
		string argumentName,
		string actualValue
	) cil managed aggressiveinlining 
{
	.custom instance void JetBrains.Annotations.AssertionMethodAttribute::.ctor() = (
		01 00 00 00
	)
	.param [1]
		.custom instance void JetBrains.Annotations.InvokerParameterNameAttribute::.ctor() = (
			01 00 00 00
		)
	.param [2]
		.custom instance void JetBrains.Annotations.AssertionConditionAttribute::.ctor(valuetype JetBrains.Annotations.AssertionConditionType) = (
			01 00 03 00 00 00 00 00
		)
	// Method begins at RVA 0x9050
	// Code size 25 (0x19)
	.maxstack 8

	IL_0000: ldarg.0
	IL_0001: ldarg.1
	IL_0002: call !!0 Remotion.Utilities.ArgumentUtility::CheckNotNull<string>(string, !!0)
	IL_0007: pop
	IL_0008: ldarg.1
	IL_0009: callvirt instance int32 [mscorlib]System.String::get_Length()
	IL_000e: brtrue.s IL_0017

	IL_0010: ldarg.0
	IL_0011: call class [mscorlib]System.ArgumentException Remotion.Utilities.ArgumentUtility::CreateArgumentEmptyException(string)
	IL_0016: throw

	IL_0017: ldarg.1
	IL_0018: ret
} // end of method ArgumentUtility::CheckNotNullOrEmpty
```

- `CheckNotEmpty`
```text
.method public hidebysig static 
	string CheckNotEmpty (
		string argumentName,
		string actualValue
	) cil managed aggressiveinlining 
{
	.custom instance void JetBrains.Annotations.AssertionMethodAttribute::.ctor() = (
		01 00 00 00
	)
	.param [1]
		.custom instance void JetBrains.Annotations.InvokerParameterNameAttribute::.ctor() = (
			01 00 00 00
		)
	// Method begins at RVA 0x910f
	// Code size 20 (0x14)
	.maxstack 8

	IL_0000: ldarg.1
	IL_0001: brfalse.s IL_0012

	IL_0003: ldarg.1
	IL_0004: callvirt instance int32 [mscorlib]System.String::get_Length()
	IL_0009: brtrue.s IL_0012

	IL_000b: ldarg.0
	IL_000c: call class [mscorlib]System.ArgumentException Remotion.Utilities.ArgumentUtility::CreateArgumentEmptyException(string)
	IL_0011: throw

	IL_0012: ldarg.1
	IL_0013: ret
} // end of method ArgumentUtility::CheckNotEmpty
```

> does it also do inlining in debug builds?

Yes, `aggressiveinlining` is also added for debug builds.

**Comment by iam-mholle** [comment]
> 
> 
> Okay, probably misleading: I meant at the usage-site, three different methods, each using a different argument check. I want to verify the inline behavior.

Won't there be no difference? I thought the JIT compiler decides this...

**Comment by MichaelKetting** [comment]
> 
> 
> > Okay, probably misleading: I meant at the usage-site, three different methods, each using a different argument check. I want to verify the inline behavior.
> 
> Won't there be no difference? I thought the JIT compiler decides this...

ASM in Jira Ticket looks good.

**Comment by iam-mholle** [comment]
Added to master with 71f0fcc32433c06cee31641c4c0403fca3eacac5.

**Pull request closed without merging**